AVRO-4026: [c++] Add new methods to CustomAttributes to allow non-string custom attributes in schemas

jhump opened the pull request:
This adds new methods to `CustomAttributes` to allow setting non-string values. These other methods work with JSON-encoded strings.

Unlike #3064 and #3069, this change attempts to be backwards compatible. However, from reading more comments in pull requests, it looks like the "fix" I added (to escape the keys and values in custom attributes when printing to JSON) may actually be a compatibility issue since it seems that users were _expected_ to have to escape string values if they contained any characters that would be escaped in JSON (including quotes). That seems like a really terrible API, and it also meant that the values would not round-trip correctly: reading a data file would not create custom attributes with these strings properly escaped, so later writing out data with the same schema would generate an invalid schema JSON document.

In any event, this uses strings as the values even though it would be ideal if we could pass some sort of structured data as the value type. The ideal types (`json::Entity` and its accompanying `json::Object` and `json::Array` types) are defined in `json/JsonDom.hh`. But that header files is not part of the Avro include files distribution, which means we cannot `#include` it from `CustomAttributes.hh`, so it's a no-go. From a little history spelunking, I see that they indeed used to use a structured form which was simplified to strings in #1821, purely because these JSON header files aren't available to users in the Avro distribution.

Alternatives that I considered for using JSON-encoded strings: 
* Include the JSON types in the Avro distribution. I think this would be as simple as moving the header files out of `lang/c++/impl` and into `lang/c++/include/avro`. But then we expand the public API of too much. This approach was already tried and rejected in #1820.
* Use a wrapper like `std::any` as the value type. This _can_ be `#include`d in `CustomAttributes.hh` but eliminates type safety in the signature. The only concrete accepted value would likely be `json::Entity` -- though we could make it more sophisticated and also allow the various value types sans wrapper: `std::string`, `bool`, `int64_t`,  `double`, `json::Array` (aka `std::vector<json::Entity>`), and `json::Object` (aka `std::map<std::string, json::Entity>`). But this isn't really usable by external/user code, at least not for any composite values, since they aren't able to include the JSON headers and then produce valid values of `json::Entity`.
* Convert from `json::Entity` to some other structured type that is defined in `CustomAttributes.hh`. This could possibly be a concrete `std::variant` that allows the various options (and could use `std::monostate` to represent JSON null values). This introduces non-trivial conversion code. From a performance perspective, it could likely be better than converting to/from strings, but it's a non-trivial amount of new-new code to maintain, which didn't feel right.

## What is the purpose of the change
Fixes exception from C++ library when compiling schemas with non-string custom attributes.

## Verifying this change
This change added tests and can be verified as follows:
* Updates existing unit test cases to use the new method and use non-string custom attributes.

## Documentation
* Does this pull request introduce a new feature? yes
* If yes, how is the feature documented? not documented

martin-g commented:
@wgtmac Do you have an opinion about these changes ?
There are two more PR about the same (https://github.com/apache/avro/pull/3064 and https://github.com/apache/avro/pull/3069). We try to find the best one of them.

wgtmac commented:
I'm not fully familiar with the context (and the code) yet. I agree that keeping backward compatibility is the bottom line. However, from the JIRA issue it seems that non-string attributes were supported once but broken after fixing other issues. Shouldn't we just fix it to revert the unexpected behavior? Except the alternatives list above, is it possible to add an option to `CustomAttributes` to control its behavior of dealing with non-string values?

jhump commented:
> Except the alternatives list above, is it possible to add an option to CustomAttributes to control its behavior of dealing with non-string values?

Most certainly. I hadn't considered it, but it would be easy to change to that. Would like you me to apply that suggestion in this PR?

jhump commented:
@wgtmac, I've updated this branch so that instead of deprecating the old methods and adding new ones, there's now a constructor flag to indicate whether values are strings or arbitrary JSON values (in which case string values must be quoted).

In updating tests, I realized that my code wasn't correctly validating that string values were correct -- if it had a botched/unescaped quote, it wouldn't be flagged as an invalid value. So I added some stuff in the json folder so that `loadEntity` makes sure that it exhausts all input and fails if there appears to be a subsequent value in the stream. This also caught an existing issue in a schema string literal in `SchemaTests.cc`.

jhump commented:
@wgtmac, if you don't like the look of this API -- using a separate constructor with a bool flag -- let me know. I'm happy to roll back that commit or continue to iterate on this.

wgtmac commented:
@jhump I'm not a maintainer so my opinion might not be thorough. 

@martin-g @thiru-mg WDYT?

jhump commented:
@wgtmac, what do you think of the current approach? If I just replace the bool constructor parameter with an enum (as mentioned above), would that be pretty close to an acceptable patch? Any other concerns or feedback?

wgtmac commented:
@jhump Yes, I think so.

jhump commented:
@wgtmac, I've updated the PR with that suggestion. Please take a look. Thanks in advance!

wgtmac reviewed (commented):
Thanks for the change! I did a quick scan on this PR and left some comments about the expected behavior. I need to spend more time on the json impl in this code base.

jhump commented on lang/c++/include/avro/CustomAttributes.hh:
```diff
@@ -19,27 +19,76 @@
+    // WILL be escaped and other special characters MAY be escaped.
+    //
+    // To support non-string values, use CustomAttributes(CustomAttributes::json) instead.
+    CustomAttributes() : CustomAttributes(string) {}
```
We always use `ValueMode::json` when we read an Avro file -- it doesn't matter how the file was created because the file always contains valid JSON in the schema. So if it was written with an older version, then it will only ever have string values in custom attributes. But that is fine to later load and read with `ValueMode::json` since that handles _any_ kind of attribute value.

This is here for backwards compatibility of _users_ of this API since this is a public type defined in a public header. So even though internal usages of `CustomAttributes` always should use `ValueMode::json`, that's not necessarily the case for external usages.

wgtmac replied:
OK, that makes sense.

jhump reviewed (commented):
@wgtmac, thanks for the review! Sorry for the delay in this reply. I haven't pushed any changes, but I will address some of the comments in my next push, later today.

wgtmac commented on lang/c++/include/avro/CustomAttributes.hh:
```diff
@@ -19,27 +19,76 @@
 public:
-    // Retrieves the custom attribute json entity for that attributeName, returns an
-    // null if the attribute doesn't exist.
+    enum ValueMode {
+        // When a CustomAttributes is created using this mode, all values are strings.
+        // The value should not be quoted, but any interior quotes and special
+        // characters (such as newlines) must be escaped.
+        string,
+        // When a CustomAttributes is created using this mode, all values are JSON
+        // values. String values must be quoted and escaped.
+        json
+    };
```
Isn't `valueMode_` not serialized in the JSON string? Upon creating the `CustomAttributes` for deserialization, we don't know the `ValueMode` used to serialize it and only set `ValueMode` we choose to the constructor.

jhump replied:
The value mode simply determines how this `CustomAttributes` interprets values provided via `setAttribute` and in turn how they are returned from `getAttributes` and the `attributes()` map accessor. It's an entirely runtime property.

We don't need to know the value mode used to serialize a JSON schema. The JSON string is known to be valid JSON at this point., since the compiler will have already parsed it into an entity. It may have only string custom attributes or it might not. If it was serialized using "json" value mode, we can still process it with "string" value mode, as long as all custom attributes are strings. And we can always process any input in "json" value mode (the new non-deprecated mode).

The current logic (prior to this PR) _cannot_ handle custom attributes unless they are strings. While this library cannot currently _generate_ schemas that have non-string custom attributes (at least as of v1.11.2), all other Avro libraries _can_ generate that sort of file. That's the bug we're trying to fix in this PR: this library currently throws an exception if an Avro file has a schema with non-string custom attributes.

wgtmac replied:
Thanks for the explanation! It sounds good to me.

jhump replied:
Updated the enum to use `enum class` and upper-case constant names in 2fd25e1

jhump commented on lang/c++/include/avro/CustomAttributes.hh:
```diff
@@ -19,27 +19,76 @@
 #define avro_CustomAttributes_hh__
 
-#include "Config.hh"
 #include <iostream>
```
I don't think so. One of the methods of `CustomAttributes` uses it:
```
void printJson(std::ostream &os, const std::string &name) const;
```

jhump commented on lang/c++/include/avro/CustomAttributes.hh:
```diff
@@ -19,27 +19,76 @@
+
+    // Creates a new CustomAttributes object.
+    //
+    // If the given mode is string, all values must be strings. All values passed to
```
Trimmed away redundant comments in 0f0efd2.

jhump commented on lang/c++/include/avro/CustomAttributes.hh:
```diff
@@ -19,27 +19,76 @@
+    // WILL be escaped and other special characters MAY be escaped.
+    //
+    // To support non-string values, use CustomAttributes(CustomAttributes::json) instead.
+    CustomAttributes() : CustomAttributes(string) {}
```
Deprecated these elements in 0f0efd2.

I am testing using clang on OS X. I don't see any other deprecations in this C++ codebase, so I can't find any other examples of muting the compiler warnings that result from using the deprecated elements in the implementation and the tests. Let me know how I should handle this. For now, I've added clang-specific `#pragma` statements to eliminate the noisy warnings.

jhump commented on lang/c++/test/unittest.cc:
```diff
@@ -489,12 +491,26 @@ struct TestSchema {
-    void checkCustomAttributes_getAttribute() {
-        CustomAttributes cf;
-        cf.addAttribute("field1", std::string("1"));
+    void checkCustomAttributes_addAndGetAttributeJson() {
```
I think the only thing missing from the matrix was the failure tests for passing a "string" value mode style value when mode is actually "json" and vice versa.

I updated the tests in e95f691, and they caught an issue: we were only handling `TooManyExceptions` when value mode was "string", but it could be triggered in either value mode. So I've fixed that in the same commit.

jhump commented:
@wgtmac, I just pushed some revisions that I _think_ address all of your feedback and include all the changes you requested. Please take another look.

wgtmac commented:
Thanks for the update! I will take a look after I'm back from vacation.

wgtmac commented on lang/c++/impl/CustomAttributes.cc:
```diff
@@ -16,26 +16,63 @@
+#if defined(__clang__)
+// Even though CustomAttributes::ValueMode::STRING is deprecated, we still have to
+// handle/implement it.
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
```
Is it better to add `#pragma clang diagnostic push|pop` to wrap only `#include "CustomAttributes.hh"`?

jhump replied:
That doesn't work here. It's the implementation in the body of this file that references the deprecated elements. Doing this only for the `#include` does nothing. If I did use `push|pop`, we'd have to do it around all three functions that have offending usages: the constructor, `addAttribute`, and `printJson`. That starts to get very verbose, and it's unclear that it's helpful: if the implementation added another method that needed to handle `ValueMode::STRING`, we'd have to wrap yet another place. So it feels appropriate to have it apply to the whole file. I can move it below to make it clear that it applies to the implementation body, not an include directive.

I **can**, however, use `push|pop` in `uinttest.cpp`, since it's only one function there that tests (and thus uses) the deprecated elements.

jhump replied:
Done

wgtmac commented on lang/c++/impl/CustomAttributes.cc:
```diff
@@ -16,26 +16,63 @@
+    const std::string &jsonVal = (valueMode_ == ValueMode::STRING)
+                ? std::move("\"" + value + "\"")
+                : value;
+    try {
```
Is it expensive to validate the input? If we don't do this, what will happen for malformed input?

jhump replied:
If we don't validate it, then we could end up generating invalid Avro files.

I suppose we could also make a private method that skips the validation and declare the Compiler as a friend class that can invoke it. From the compiler, the values will have already been validated when the JSON schema was originally parsed, so it would be fine to skip in those cases.

The existing "string only" usage is error-prone enough, since it still requires that interior quotes and newlines be escaped (since all it does is add a leading and trailing `"` to turn it into a JSON value). But I figured it would be even more error-prone with allowing any arbitrary JSON element (including complicated values like arrays and objects) and also more error-prone now that it has a modality that is defined at construction time. Failing fast seemed like a much better developer experience than ignoring such issues and letting the library generate an invalid file.

jhump replied:
> I suppose we could also make a private method that skips the validation and declare the Compiler as a friend class that can invoke it.

Welp, this won't work. The compiler isn't a class but a set of functions. But the real issue is that the `CustomAttributes.addAttribute` method is called from a static function in `Compiler.cc, so we can't possibly make it a friend of `CustomAttributes`. If you have any other recommendation for how we might omit the validation just from the compiler, I will happily implement it.

I suppose another option is to just make it another constructor argument, like `ValidationMode`, and then document that it's unwise to use the "no validate" mode unless the caller is doing something else to validate the values before providing them, since it could otherwise result in invalid/unreadable Avro files.

What do you think?

wgtmac replied:
That may introduce a larger change so I think we need opinions from the maintainers. Let's keep it simple in this PR.

jhump replied:
The other approach is to simply remove validation entirely. But I feel bad doing that seems it makes the API a foot-gun: it becomes easy to accidentally produce an Avro file that is junk/unreadable, even though the function calls that produced the file never returned any errors or threw any exceptions.

FWIW, in the _expected_ cases, where custom values are not present in great number and have values that are simple scalars and short strings, it should be very cheap relative to the rest of the process of encoding and creating the Avro file. For the compiler, reading and processing an Avro file, the extra overhead would be strictly less than the processing needed to initially parse the JSON schema (likely _much_ less, since custom attributes are typically a very small portion of the JSON). And I suspect (though have no profile data) that parsing the JSON schema is already a fairly light-weight part of reading an Avro file.

@martin-g, thoughts?

wgtmac commented on lang/c++/impl/json/JsonIO.cc:
```diff
@@ -49,6 +49,28 @@ char JsonParser::next() {
 }
 
+bool JsonParser::hasMore() {
+    if (peeked || hasNext && !isspace(nextChar)) {
```
```suggestion
    if (peeked || (hasNext && !isspace(nextChar))) {
```

jhump replied:
Fixed

wgtmac commented on lang/c++/impl/json/JsonIO.hh:
```diff
@@ -116,6 +116,8 @@ public:
         return curToken;
     }
 
+    bool hasMore();
```
Is it possible not to add `hasMore()` by reusing `peek()` above?

jhump replied:
It is possible. But the reason I did not do that is because `peek()` throws an exception if it's at the end, so it seemed better to add a way to check for more instead of trying to peek and catching. Also the current exception is just an `Exception`, not something like `EOFException`, so the code would need to actually inspect the string message to make sure it is EOF, which seemed brittle and undesirable.

jhump commented on lang/c++/include/avro/CustomAttributes.hh:
```diff
@@ -19,27 +19,62 @@
+        // to be strings. The value should not be quoted, but any interior quotes and
+        // special characters (such as newlines) must be escaped.
+        STRING
+        [[deprecated("The JSON ValueMode is less error-prone and less limited.")]],
```
I have no idea what to put for "XXX" version. Is there a standard support policy for deprecating things and making backwards-breaking changes like this?

martin-g replied:
Deprecated since 1.13.0
To be removed with 1.14.0

This is the semver followed by Avro. Here the leading number (`1`) is the version of the Avro Schema

jhump replied:
I've added some info about the versions to the deprecation messages, but it could probably use some tweaking, particularly with what to say about the no-arg constructor and how that should be handled.

jhump commented on lang/c++/impl/CustomAttributes.cc:
```diff
@@ -16,26 +16,63 @@
+    // from a header file: CustomAttributes.hh. But the json header files are not
+    // part of the Avro distribution (intentionally), so CustomAttributes.hh cannot
+    // #include any of the json header files.
+    const std::string &jsonVal = (valueMode_ == ValueMode::STRING)
+                ? std::move("\"" + value + "\"")
+                : value;
```
My bad. I had hoped I could avoid a copy with `std::move`, but I see now that I'm using it wrong.

jhump replied:
Fixed

jhump commented:
@wgtmac, @martin-g, @KalleOlaviNiemitalo, I addressed all of your comments I think (or at least replied to them all, in some cases with additional clarifying questions). Anything else I can do to help move this along?

wgtmac reviewed (approved):
LGTM. Thanks!

jhump commented:
@martin-g, anything else I can do to help get this change merged?

@thiru-mg, I see you recently approved and merged a related fix in #3308. Any chance you could help review this?

martin-g commented:
> @martin-g, anything else I can do to help get this change merged?

I am not a C++ developer, so I cannot tell whether the changes are good or not.
Let's wait for @thiru-mg to review it!
Otherwise if there are at least two approvals by community member I could help with the merge too!

wgtmac commented:
@thiru-mg @Fokko Could you help review this? 

I have been integrating avro-cpp with iceberg and this issue blocks me to set `field-id` and other custom attributes to Iceberg schema.

jhump commented:
Superceded by #3344
